### PR TITLE
First implementation of CASProxy

### PR DESCRIPTION
**Pull request by vmax**

### CODE REVIEW

File: tools/reviewer/cas_proxy/ProxyClientTool.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    StreamObserver<AuthProtos.FileUploadRequest> req =
+        nonBlockingStub.uploadFile(
+            new StreamObserver<AuthProtos.FileUploadResponse>() {
+              @Override
+              public void onNext(AuthProtos.FileUploadResponse value) {
+                System.err.println(value.toString());
+              }
+
+              @Override
+              public void onError(java.lang.Throwable t) {
```
Comment (oferb) [inline tools/reviewer/cas_proxy/ProxyClientTool.java]: `t-<throwable`

File: tools/reviewer/cas_proxy/ProxyClientTool.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+import io.grpc.stub.StreamObserver;
+import java.io.File;
+import java.io.FileInputStream;
+import java.nio.file.Paths;
+import java.util.Iterator;
+import java.util.concurrent.Executors;
+
+public class ProxyClientTool {
+
+  public static void uploadFile(CASProxyServiceStub nonBlockingStub, String fileName)
```
Comment (oferb) [inline tools/reviewer/cas_proxy/ProxyClientTool.java]: `filename->absFilename`

File: tools/reviewer/cas_proxy/Proxy.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class Proxy {
+
+  // TODO(vmax): make bucket configurable
+  public static final String CAS_BUCKET_NAME = "test-bucket-cas";
+
+  static class AccessDeniedException extends Exception {}
+
+  static class Service extends CASProxyServiceGrpc.CASProxyServiceImplBase {
```
Comment (oferb) [inline tools/reviewer/cas_proxy/Proxy.java]: Let's rename `Service->CasService`.

File: tools/reviewer/cas_proxy/Proxy.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+      }
+    }
+
+    @Override
+    public StreamObserver<FileUploadRequest> uploadFile(
+        StreamObserver<FileUploadResponse> responseObserver) {
+      return new FileUploader(this.client, this.accessManager, responseObserver);
+    }
+
+    public Service(String serviceAccountPath, AccessManager accessManager) {
```
Comment (oferb) [inline tools/reviewer/cas_proxy/Proxy.java]: Can you put the ctors on top?

File: tools/reviewer/cas_proxy/Proxy.java [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+      this.client = new FirestoreProtoClient(serviceAccountPath);
+      if (accessManager == null) {
+        this.accessManager = new PublicAccess();
+      } else {
+        this.accessManager = accessManager;
+      }
+    }
+
+    public Service(String serviceAccountPath) {
+      this(serviceAccountPath, null);
```
Comment (oferb) [inline tools/reviewer/cas_proxy/Proxy.java]: I think it's cleaner to `null->new PublicAccess()` and not handle null in `Service(String serviceAccountPath, AccessManager accessManager)`.

Probably even better though is to just have 1 constructor and pass the AccessManager you want. Default to public is not a good practice :)

File: tools/reviewer/cas_proxy/Proxy.java [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+      }
+
+      @Override
+      public void onNext(FileUploadRequest value) {
+        if (this.accessManager.hasWriteAccess(value.getIdToken())) {
+          hasAccess.set(true);
+        }
+
+        if (!hasAccess.get()) {
+          responseObserver.onError(new AccessDeniedException());
```
Comment (oferb) [inline tools/reviewer/cas_proxy/Proxy.java]: Finally, using `AccessDeniedException`! :)

File: tools/reviewer/cas_proxy/Proxy.java [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+  public static final String CAS_BUCKET_NAME = "test-bucket-cas";
+
+  static class AccessDeniedException extends Exception {}
+
+  static class Service extends CASProxyServiceGrpc.CASProxyServiceImplBase {
+    private FirestoreProtoClient client;
+    private AccessManager accessManager;
+    private HashMap<String, String> fileCache = new HashMap<>();
+
+    static class FileUploader implements StreamObserver<FileUploadRequest> {
```
Comment (oferb) [inline tools/reviewer/cas_proxy/Proxy.java]: `FileUploader` also downloads file?

File: tools/reviewer/cas_proxy/Proxy.java [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+        responseObserver.onError(t);
+      }
+
+      @Override
+      public void onCompleted() {
+        String sha256 = Hashing.sha256().hashBytes(buffer.toByteArray()).toString();
+        File tempFile;
+        try {
+          tempFile = File.createTempFile(sha256, ".tmp");
+          FileOutputStream outputStream = new FileOutputStream(tempFile);
```
Comment (oferb) [inline tools/reviewer/cas_proxy/Proxy.java]: You should probably try-with-resources:
https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

File: tools/reviewer/cas_proxy/Proxy.java [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+      }
+
+      @Override
+      public void onError(Throwable t) {
+        responseObserver.onError(t);
+      }
+
+      @Override
+      public void onCompleted() {
+        String sha256 = Hashing.sha256().hashBytes(buffer.toByteArray()).toString();
```
Comment (oferb) [inline tools/reviewer/cas_proxy/Proxy.java]: So we keep everything in the buffer, in memory?
Why not write it to file on `onNext()` and then close it on `onCompleted()`?

File: tools/reviewer/cas_proxy/Proxy.java [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+        File tempFile;
+        try {
+          tempFile = File.createTempFile(sha256, ".tmp");
```
Comment (oferb) [inline tools/reviewer/cas_proxy/Proxy.java]: If someone uploads and then downloads, we can already keep it in the local disk, no?

File: tools/reviewer/cas_proxy/Proxy.java [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+      if (!this.accessManager.hasReadAccess(request.getIdToken(), request.getSha256())) {
+        responseObserver.onError(new AccessDeniedException());
+      }
+
+      String fileName =
+          fileCache.computeIfAbsent(
+              request.getSha256(),
+              (sha256) -> {
+                try {
+                  return client.downloadFrom(CAS_BUCKET_NAME, sha256);
```
Comment (oferb) [inline tools/reviewer/cas_proxy/Proxy.java]: Where does it download to? Where is the cache located? :)

File: tools/reviewer/cas_proxy/PublicAccess.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.startupos.tools.reviewer.cas_proxy;
+
+public class PublicAccess implements AccessManager {
+
+  @Override
+  public boolean hasReadAccess(String token, String sha256) {
```
[the remark this answers is not in the document]
Comment (vmax) [reply tools/reviewer/cas_proxy/PublicAccess.java]: @oferb do you mean `TODO` of implementing something other than `PublicAccess` or `TODO` of having `hasDeleteAccess` in interface?

File: tools/reviewer/cas_proxy/ProxyClientTool.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    if (!args[0].equals("upload") && !args[0].equals("download")) {
+      System.err.println("Error: should pass <download|upload> as first argument");
+      System.exit(1);
+    }
+
+    if (args[0].equals("upload")) {
+      uploadFile(nonBlockingStub.withExecutor(Executors.newSingleThreadExecutor()), args[1]);
+      Thread.sleep(Long.MAX_VALUE); // sleep forever to let non-blocking stub get a response
```
[the remark this answers is not in the document]
Comment (vmax) [reply tools/reviewer/cas_proxy/ProxyClientTool.java]: @oferb Yes. This is only meant for tests though, so we don't need perfect UX.

File: tools/reviewer/cas_proxy/Proxy.java (hunk 7 above)
Comment (vmax) [reply tools/reviewer/cas_proxy/Proxy.java]: @oferb nope, it is only used in `uploadFile`. From server's point of view user is _uploading_ 🙂

File: tools/reviewer/cas_proxy/Proxy.java (hunk 6 above)
Comment (vmax) [reply tools/reviewer/cas_proxy/Proxy.java]: @oferb Yep 🙂 I think custom exception class is a-ok choice here?

File: tools/reviewer/cas_proxy/Proxy.java (hunk 11 above)
Comment (vmax) [reply tools/reviewer/cas_proxy/Proxy.java]: @oferb to a temporary directory OS provides to us (you can get the path by `System.getProperty("java.io.tmpdir")`. We should probably make it configurable on next iteration.

File: tools/reviewer/cas_proxy/Proxy.java (hunk 10 above)
Comment (vmax) [reply tools/reviewer/cas_proxy/Proxy.java]: @oferb Right. I'll add a `TODO` to handle this.

File: tools/reviewer/cas_proxy/Proxy.java (hunk 9 above)
Comment (vmax) [reply tools/reviewer/cas_proxy/Proxy.java]: @oferb Right! I'll add a `TODO` and handle in new iteration

File: tools/reviewer/cas_proxy/Proxy.java (hunk 5 above)
Comment (vmax) [reply tools/reviewer/cas_proxy/Proxy.java]: @oferb Good point!

File: tools/reviewer/cas_proxy/api.proto [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+package com.google.startupos.tools.reviewer.proxy;
+
+service CASProxyService {
+  rpc downloadFile(FileDownloadRequest) returns (stream FileDownloadResponse);
+  rpc uploadFile(stream FileUploadRequest) returns (FileUploadResponse);
+}
+
+message FileDownloadRequest {
+  string idToken = 1;
```
[the remark this answers is not in the document]
Comment (vmax) [reply tools/reviewer/cas_proxy/api.proto]: @oferb fixed in https://github.com/google/startup-os/pull/541/commits/881572de2fce8e235dd42d5a74e171b80f38636c

File: tools/reviewer/cas_proxy/api.proto [hunk 15]
```diff
@@ -0,0 +1,26 @@
+syntax = "proto3";
+
+option java_package = "com.google.startupos.tools.reviewer.proxy";
```
[the remark this answers is not in the document]
Comment (vmax) [reply tools/reviewer/cas_proxy/api.proto]: @oferb fixed in https://github.com/google/startup-os/pull/541/commits/881572de2fce8e235dd42d5a74e171b80f38636c

File: tools/reviewer/cas_proxy/ProxyClientTool.java [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+    while (responseIterator.hasNext()) {
+      AuthProtos.FileDownloadResponse resp = responseIterator.next();
+      System.err.println(resp);
+      System.err.println(resp.getContent().toStringUtf8());
+    }
+  }
+
+  public static void main(String[] args) throws Exception {
+    ManagedChannel channel =
+        ManagedChannelBuilder.forAddress("localhost", 6000).usePlaintext().build();
```
[the remark this answers is not in the document]
Comment (vmax) [reply tools/reviewer/cas_proxy/ProxyClientTool.java]: @oferb fixed in https://github.com/google/startup-os/pull/541/commits/881572de2fce8e235dd42d5a74e171b80f38636c

File: tools/reviewer/cas_proxy/ProxyClientTool.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (vmax) [reply tools/reviewer/cas_proxy/ProxyClientTool.java]: @oferb fixed in https://github.com/google/startup-os/pull/541/commits/881572de2fce8e235dd42d5a74e171b80f38636c

File: tools/reviewer/cas_proxy/ProxyClientTool.java [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+                t.printStackTrace();
+              }
+
+              @Override
+              public void onCompleted() {
+                System.err.println("completed");
+              }
+            });
+
+    File upl = Paths.get(fileName).toFile();
```
[the remark this answers is not in the document]
Comment (vmax) [reply tools/reviewer/cas_proxy/ProxyClientTool.java]: @oferb fixed in https://github.com/google/startup-os/pull/541/commits/881572de2fce8e235dd42d5a74e171b80f38636c

File: tools/reviewer/cas_proxy/ProxyClientTool.java [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+    FileInputStream stream = new FileInputStream(upl);
+
+    byte[] chunk = new byte[1024];
```
[the remark this answers is not in the document]
Comment (vmax) [reply tools/reviewer/cas_proxy/ProxyClientTool.java]: @oferb fixed in https://github.com/google/startup-os/pull/541/commits/881572de2fce8e235dd42d5a74e171b80f38636c

File: tools/reviewer/cas_proxy/ProxyClientTool.java (hunk 1 above)
Comment (vmax) [reply tools/reviewer/cas_proxy/ProxyClientTool.java]: @oferb fixed in https://github.com/google/startup-os/pull/541/commits/881572de2fce8e235dd42d5a74e171b80f38636c

File: tools/reviewer/cas_proxy/ProxyClientTool.java (hunk 2 above)
Comment (vmax) [reply tools/reviewer/cas_proxy/ProxyClientTool.java]: @oferb fixed in https://github.com/google/startup-os/pull/541/commits/881572de2fce8e235dd42d5a74e171b80f38636c

File: tools/reviewer/cas_proxy/Proxy.java (hunk 3 above)
Comment (vmax) [reply tools/reviewer/cas_proxy/Proxy.java]: @oferb fixed in https://github.com/google/startup-os/pull/541/commits/881572de2fce8e235dd42d5a74e171b80f38636c

File: tools/reviewer/cas_proxy/Proxy.java (hunk 4 above)
Comment (vmax) [reply tools/reviewer/cas_proxy/Proxy.java]: @oferb fixed in https://github.com/google/startup-os/pull/541/commits/881572de2fce8e235dd42d5a74e171b80f38636c

File: tools/reviewer/cas_proxy/Proxy.java (hunk 5 above)
Comment (vmax) [reply tools/reviewer/cas_proxy/Proxy.java]: @oferb fixed in https://github.com/google/startup-os/pull/541/commits/881572de2fce8e235dd42d5a74e171b80f38636c

File: tools/reviewer/cas_proxy/Proxy.java (hunk 8 above)
Comment (vmax) [reply tools/reviewer/cas_proxy/Proxy.java]: @oferb fixed in https://github.com/google/startup-os/pull/541/commits/881572de2fce8e235dd42d5a74e171b80f38636c

File: tools/reviewer/cas_proxy/Proxy.java (hunk 10 above)
Comment (vmax) [reply tools/reviewer/cas_proxy/Proxy.java]: @oferb `todo` in https://github.com/google/startup-os/pull/541/commits/d751e511a2d7fe83561daa67f717aa702dc3da71

File: tools/reviewer/cas_proxy/Proxy.java (hunk 9 above)
Comment (vmax) [reply tools/reviewer/cas_proxy/Proxy.java]: @oferb `todo` in https://github.com/google/startup-os/pull/541/commits/d751e511a2d7fe83561daa67f717aa702dc3da71

File: tools/reviewer/cas_proxy/PublicAccess.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (elinabtl) [reply tools/reviewer/cas_proxy/PublicAccess.java]: Implementing something other than `PublicAccess` :)

[the remark this answers is not in the document]
Comment (oferb) [reply tools/reviewer/cas_proxy/PublicAccess.java]: Something other than `PublicAccess`.

File: tools/reviewer/cas_proxy/Proxy.java (hunk 7 above)
Comment (oferb) [reply tools/reviewer/cas_proxy/Proxy.java]: Got it :)

File: tools/reviewer/cas_proxy/Proxy.java (hunk 11 above)
Comment (oferb) [reply tools/reviewer/cas_proxy/Proxy.java]: Let's add a TODO for it..

**Comment by vmax** [comment]
This will be moved to [hasadna/hasadna](https://github.com/hasadna/hasadna)

**Pull request closed without merging**